core/search/disjunction.hpp

```diff
@@ -1314,60 +1301,128 @@ template<typename DocIterator, typename Merger,
+template<typename Disjunction, typename Merger, typename... Args>
+doc_iterator::ptr MakeDisjunction(typename Disjunction::doc_iterators_t&& itrs,
+                                  Merger&& merger, Args&&... args) {
   const auto size = itrs.size();
 
-  switch (size) {
-    case 0:
-      // empty or unreachable search criteria
-      return doc_iterator::empty();
-    case 1:
-      if constexpr (Disjunction::enable_unary()) {
-        using unary_disjunction_t = typename Disjunction::unary_disjunction_t;
-        return memory::make_managed<unary_disjunction_t>(
-            std::move(itrs.front()));
-      }
+  if (0 == size) {
+    // Empty or unreachable search criteria
+    return doc_iterator::empty();
+  }
 
-      // single sub-query
-      return std::move(itrs.front());
-    case 2: {
-      using basic_disjunction_t = typename Disjunction::basic_disjunction_t;
+  if (1 == size) {
+    // Single sub-query
+    if constexpr (Disjunction::kEnableUnary) {
+      using UnaryDisjunction = typename RebindIterator<Disjunction>::Unary;
 
-      // simple disjunction
-      return memory::make_managed<basic_disjunction_t>(
-          std::move(itrs.front()), std::move(itrs.back()),
-          std::forward<Args>(args)...);
+      return memory::make_managed<UnaryDisjunction>(std::move(itrs.front()));
     }
+
+    return std::move(itrs.front());
   }
 
-  if constexpr (Disjunction::small_disjunction_upper_bound() > 0) {
-    if (size <= Disjunction::small_disjunction_upper_bound()) {
-      using small_disjunction_t = typename Disjunction::small_disjunction_t;
+  if (2 == size) {
+    // 2-way disjunction
+    using BasicDisjunction = typename RebindIterator<Disjunction>::Basic;
+
+    return memory::make_managed<BasicDisjunction>(
+        std::move(itrs.front()), std::move(itrs.back()),
+        std::forward<Merger>(merger), std::forward<Args>(args)...);
+  }
 
-      // small disjunction
-      return memory::make_managed<small_disjunction_t>(
-          std::move(itrs), std::forward<Args>(args)...);
+  if constexpr (Disjunction::kSmallDisjunctionUpperBound > 0) {
+    if (size <= Disjunction::kSmallDisjunctionUpperBound) {
+      using SmallDisjunction = typename RebindIterator<Disjunction>::Small;
+
+      return memory::make_managed<SmallDisjunction>(
+          std::move(itrs), std::forward<Merger>(merger),
+          std::forward<Args>(args)...);
     }
   }
 
-  // disjunction
   return memory::make_managed<Disjunction>(std::move(itrs),
+                                           std::forward<Merger>(merger),
                                            std::forward<Args>(args)...);
 }
 
+// Returns weak conjunction iterator created from the specified sub iterators
+template<typename WeakConjunction, typename Merger, typename... Args>
+doc_iterator::ptr MakeWeakDisjunction(
+    typename WeakConjunction::doc_iterators_t&& itrs, size_t min_match,
+    Merger&& merger, Args&&... args) {
+  // FIXME(gnusi)
```

Comment:
Why don't you want make it? Maybe some comment?

Reply:
left a comment